libgraph/include/katana/TransformationView.h

```diff
@@ -0,0 +1,105 @@
+/// analytics routines as a regular property graph.
+class KATANA_EXPORT TransformationView final : public PropertyGraph {
+public:
+  TransformationView() = default;
+  TransformationView(TransformationView&& other) = default;
+
+  virtual ~TransformationView();
+
+  /// Make a projected graph from a property graph. Shares state with
+  /// the original graph.
+  static std::unique_ptr<TransformationView> MakeProjectedGraph(
+      const PropertyGraph& pg, const std::vector<std::string>& node_types,
+      const std::vector<std::string>& edge_types);
+
+  /// Return the number of nodes of the original property graph.
+  uint64_t NumOriginalNodes() const final {
+    return transformation_.original_to_transformed_nodes_.size();
+  }
+
+  /// Return the number of edges of the original property graph.
+  uint64_t NumOriginalEdges() const final {
+    return transformation_.original_to_transformed_edges_.size();
+  }
+
+  /// @param eid the input eid (must be projected edge id)
+  Edge TransformedToOriginalEdgeId(const Edge& eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < NumEdges());
+    return topology().GetLocalEdgeIDFromOutEdge(eid);
+  }
+
+  /// @param nid the input node id (must be projected node id)
+  Node TransformedToOriginalNodeId(const Node& nid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(nid < NumNodes());
+    return topology().GetLocalNodeID(nid);
+  }
+
+  /// @param eid the input eid (must be original edge id)
+  Edge OriginalToTransformedEdgeId(const Edge& eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < NumEdges());
+    return transformation_.original_to_transformed_edges_[eid];
+  }
+
+  /// @param nid the input node id (must be original node id)
+  Node OriginalToTransformedNodeId(const Node& nid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(nid < NumNodes());
+    return transformation_.original_to_transformed_nodes_[nid];
+  }
+
+  /// Bitmask of nodes included in the transformation view. Should be used to construct arrow tables.
+  std::shared_ptr<arrow::Buffer> NodeBitmask() const noexcept {
+    auto length = transformation_.original_to_transformed_nodes_.size();
+    return std::make_shared<arrow::Buffer>(
+        transformation_.node_bitmask_data_.data(),
+        arrow::BitUtil::BytesForBits(length));
+  }
+
+  /// Bitmask of edges included in the transformation view. Should be used to construct arrow tables.
+  std::shared_ptr<arrow::Buffer> EdgeBitmask() const noexcept {
+    auto length = transformation_.original_to_transformed_edges_.size();
+    return std::make_shared<arrow::Buffer>(
+        transformation_.edge_bitmask_data_.data(),
+        arrow::BitUtil::BytesForBits(length));
+  }
+
+private:
+  struct Transformation {
+    NUMAArray<Node> original_to_transformed_nodes_;
+    NUMAArray<Edge> original_to_transformed_edges_;
+    NUMAArray<uint8_t> node_bitmask_data_;
+    NUMAArray<uint8_t> edge_bitmask_data_;
+  };
```

Comment:
Why bundle this up?

Reply:
No reason besides personal taste. I don't feel strongly about it, though.

Reply:
I feel like it creates a level of "I gotta look somewhere else to understand this" to a number of constructors. But I also don't have a strong opinion. C++ code is hard to follow no matter what you do, so things like this are not that important.

Reply:
I would slightly prefer making these private members of the `TransformationView`